src/arch/cortex-m0/ctx_switch.S

```diff
@@ -27,29 +27,68 @@ EXC_RETURN_PSP:
   .word 0xFFFFFFFD
 
 .thumb_func
-/* r0 is top of user stack, r1 is heap base */
+/* r0 is top of user stack, r1 Process GOT */
 switch_to_user:
-  /* Load bottom of stack into Process Stack Pointer */
-  msr psp, r0
-
-  /* Cortex-M0 can only push registers R0-R7 directly, so move R8-R11 to R0-R3.
-   * This is equivalent to the 32-bit "push {r4-r11}" instruction. */
-  push {r4-r7}
-  mov r4,  r8
-  mov r5,  r9
-  mov r6, r10
-  mov r7, r11
-  push {r4-r7}
-
-  mov r9, r1
-  svc 0xff
-
-  /* These instructions are equivalent to the 32-bit "pop {r4-r11}" */
-  pop {r4-r7}
-  mov  r8, r4
-  mov  r9, r5
-  mov r10, r6
-  mov r11, r7
-  pop {r4-r7}
-
-  bx lr
+    /* Cortex-M0 can only push registers R0-R7 directly, so move R8-R11 to R0-R3.
+     * This is equivalent to the 32-bit "push {r4-r11}" instruction. */
+    push {r4-r7}
+    mov r4,  r8
+    mov r5,  r9
+    mov r6, r10
+    mov r7, r11
+    push {r4-r7}
+
+    /* Load non-hardware-stacked registers from Process stack */
+    subs r0, #32
+    ldr r7, [r0, #0]
+    ldr r6, [r0, #4]
+    ldr r5, [r0, #8]
+    ldr r4, [r0, #12]
```

Comment:
Just did, looks pretty good, a few small things noted inline.

On Mon, Jul 11, 2016 at 10:55 AM Amit Levy <notifications@github.com> wrote:

> In src/arch/cortex-m0/ctx_switch.S
> <https://github.com/helena-project/tock/pull/61#discussion_r70273291>:
>
> > -  bx lr
> > +    /* Cortex-M0 can only push registers R0-R7 directly, so move R8-R11 to R0-R3.
> > +     * This is equivalent to the 32-bit "push {r4-r11}" instruction. */
> > +    push {r4-r7}
> > +    mov r4,  r8
> > +    mov r5,  r9
> > +    mov r6, r10
> > +    mov r7, r11
> > +    push {r4-r7}
> > +
> > +    /* Load non-hardware-stacked registers from Process stack */
> > +    subs r0, #32
> > +    ldr r7, [r0, #0]
> > +    ldr r6, [r0, #4]
> > +    ldr r5, [r0, #8]
> > +    ldr r4, [r0, #12]
>
> Fixed, can you take sanity checking pass on my implementation?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/helena-project/tock/pull/61/files/6ff9e6e66273d3ef1187097dd26f8d1ec1192b7a#r70273291>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAUt3iuDDYpknkElcGUX10J5Vm4VNs5Zks5qUllBgaJpZM4JIJ1t>
> .
>